sys/net/gnrc/netif/gnrc_netif.c

```diff
@@ -1298,6 +1298,107 @@ int gnrc_netif_ipv6_add_prefix(gnrc_netif_t *netif,
 out:
     return res;
 }
+
+#if IS_USED(MODULE_GNRC_NETIF_BUS)
+static bool _has_global_addr(gnrc_netif_t *netif)
+{
+    bool has_global = false;
+
+    gnrc_netif_acquire(netif);
+
+    for (unsigned i = 0; i < CONFIG_GNRC_NETIF_IPV6_ADDRS_NUMOF; i++) {
+        if (ipv6_addr_is_unspecified(&netif->ipv6.addrs[i])) {
+            continue;
+        }
+        if (!ipv6_addr_is_link_local(&netif->ipv6.addrs[i])) {
+            has_global = true;
+            break;
+        }
+    }
+
+    gnrc_netif_release(netif);
+    return has_global;
+}
+
+bool gnrc_netif_ipv6_wait_for_global_address(gnrc_netif_t *netif,
+                                             uint32_t timeout_ms)
+{
+    msg_bus_entry_t subs[gnrc_netif_numof()];
+    unsigned count = 0;
+    bool has_global = false;
+
+    if (netif) {
+        if (_has_global_addr(netif)) {
+            return true;
+        }
+
+        msg_bus_t *bus = gnrc_netif_get_bus(netif, GNRC_NETIF_BUS_IPV6);
+        msg_bus_attach(bus, &subs[count]);
+        msg_bus_subscribe(&subs[count], GNRC_IPV6_EVENT_ADDR_VALID);
+
+    } else {
+
+        /* subscribe to all interfaces */
+        while ((netif = gnrc_netif_iter(netif))) {
+            msg_bus_t *bus = gnrc_netif_get_bus(netif, GNRC_NETIF_BUS_IPV6);
+
+            if (_has_global_addr(netif)) {
+                has_global = true;
+            }
+
+            msg_bus_attach(bus, &subs[count]);
+            msg_bus_subscribe(&subs[count], GNRC_IPV6_EVENT_ADDR_VALID);
+
+            ++count;
+        }
+
+        /* no interfaces */
+        if (count == 0) {
+            return false;
+        }
+    }
+
+    /* don't wait if we have an address already */
+    if (has_global) {
+        goto out;
+    }
```

Comment:
`!has_global` could be the `while`-condition of the loop below and spare us a `goto` ;-).